src/python/pants/backend/python/pip_requirement.py

```diff
@@ -21,8 +23,43 @@ class PipRequirement:
     """
 
     @classmethod
-    def parse(cls, line: str) -> PipRequirement:
-        return cls(pkg_resources.Requirement.parse(line))
+    def parse(cls, line: str, description_of_origin: str = "") -> PipRequirement:
+        try:
+            return cls(pkg_resources.Requirement.parse(line))
+        except InvalidRequirement as e:
+            scheme, netloc, path, query, fragment = urllib.parse.urlsplit(line, scheme="file")
+            if fragment:
+                # Try converting a pip VCS-style requirement into a PEP-440 one that can be
+                # parsed as a Requirement. E.g.,
+                # git+https://github.com/django/django.git@stable/2.1.x#egg=Django
+                # into
+                # Django@ git+https://github.com/django/django.git@stable/2.1.x#egg=Django
+
+                # Note: In pip VCS urls the fragment is a query-style string.
+                fragment_params = urllib.parse.parse_qs(fragment)
+                egg = fragment_params.get("egg")
+                if egg:
+                    # parse_qs() ignores params with empty values by default, so we're guaranteed
+                    # that there is at least one value in this list.
+                    project = egg[0]
+                    # We recompose the URL to force the default file:// scheme to be explicit.
+                    full_url = urllib.parse.urlunsplit((scheme, netloc, path, query, fragment))
+                    pep_440_req_str = f"{project}@ {full_url}"
+                    try:
+                        ret = cls(pkg_resources.Requirement.parse(pep_440_req_str))
+                        logger.warning(
```

Comment:
This is unfortunate. I asked seperately, but why are we getting in the way here still - if even just in a naggy way with a warning? My only guess is we ~need the project name parsed out for dep inference.

Reply:
I think that is exactly it. I can remove the warning though.

Reply:
Yeah I don't think that this warning is good. I'm excited about being able to remove the section in the docs explaining how you must use pep 440.

Reply:
Oh - you definitely can't remove that because of this change. As pointed out in the comments there are plenty of valid Pip requirements this does not handle. Like `.` (Pex uses this to package itself) and more. We're still getting in the way here and should just be more lax and try to get a project name and if we can't, move on.

Reply:
We also need some string to use as the target name, but we can synthesize one